File: packages/nxuskit-go/pipeline/definition.go
// Package pipeline provides pipeline definition types.
package pipeline

import (
	"encoding/json"
	"time"
)

// StageType discriminates the kind of processing a stage performs.
type StageType string

const (
	// StageTypeLLM is an LLM-only stage (calls an LLM provider).
	StageTypeLLM StageType = "llm"
	// StageTypeClipsEval runs CLIPS rules for evaluation.
	StageTypeClipsEval StageType = "clips_eval"
	// StageTypeClipsGen generates facts from CLIPS.
	StageTypeClipsGen StageType = "clips_gen"
	// StageTypeHybrid combines LLM and CLIPS processing.
	StageTypeHybrid StageType = "hybrid"
)

// RulesSource indicates where CLIPS rules come from.
type RulesSource string

const (
	// RulesSourceInline means rules are provided inline in the configuration.
	RulesSourceInline RulesSource = "inline"
	// RulesSourceFile means rules are loaded from a file path.
	RulesSourceFile RulesSource = "file"
	// RulesSourceDynamic means rules are generated at runtime.
	RulesSourceDynamic RulesSource = "dynamic"
)

// ValidationSeverity controls how dangerous CLIPS constructs are handled.
type ValidationSeverity string

const (
	// ValidationSeverityError rejects rules with dangerous constructs (default).
	ValidationSeverityError ValidationSeverity = "error"
	// ValidationSeverityWarning logs a warning but proceeds with loading.
	ValidationSeverityWarning ValidationSeverity = "warning"
	// ValidationSeverityInfo logs an info message only.
	ValidationSeverityInfo ValidationSeverity = "info"
	// ValidationSeverityIgnore skips validation entirely.
	ValidationSeverityIgnore ValidationSeverity = "ignore"
)

// ClipsConfig holds configuration for CLIPS-based stages.
type ClipsConfig struct {
	// RulesSource indicates where rules come from.
	RulesSource RulesSource `json:"rules_source" yaml:"rules_source"`

	// RulesContent contains inline rules or file path (based on RulesSource).
	RulesContent string `json:"rules_content,omitempty" yaml:"rules_content,omitempty"`

	// SaveRules indicates whether to persist rules after stage execution.
	SaveRules bool `json:"save_rules,omitempty" yaml:"save_rules,omitempty"`

	// SavePath is the path for saved rules (required if SaveRules is true).
	SavePath string `json:"save_path,omitempty" yaml:"save_path,omitempty"`

	// BSaveBinary uses CLIPS binary format for save.
	BSaveBinary bool `json:"bsave_binary,omitempty" yaml:"bsave_binary,omitempty"`

	// ValidationSeverity is the security validation severity level.
	ValidationSeverity ValidationSeverity `json:"validation_severity,omitempty" yaml:"validation_severity,omitempty"`

	// TimeoutMs is the CLIPS execution timeout in milliseconds.
	TimeoutMs *uint64 `json:"timeout_ms,omitempty" yaml:"timeout_ms,omitempty"`
}

// LlmStageConfig holds configuration for LLM-based stages.
type LlmStageConfig struct {
	// Provider name (claude, openai, ollama, etc.).
	Provider string `json:"provider" yaml:"provider"`

	// Model identifier.
	Model string `json:"model" yaml:"model"`

	// SystemPrompt is the system prompt template.
	SystemPrompt string `json:"system_prompt,omitempty" yaml:"system_prompt,omitempty"`

	// UserPrompt is the user prompt template with {{variable}} placeholders.
	UserPrompt string `json:"user_prompt" yaml:"user_prompt"`

	// Temperature is the sampling temperature.
	Temperature *float64 `json:"temperature,omitempty" yaml:"temperature,omitempty"`

	// MaxTokens is the maximum response tokens.
	MaxTokens *uint32 `json:"max_tokens,omitempty" yaml:"max_tokens,omitempty"`

	// AdditionalParams contains provider-specific parameters.
	AdditionalParams map[string]any `json:"additional_params,omitempty" yaml:"additional_params,omitempty"`

	// Providers lists multiple providers for parallel execution (optional).
	Providers []string `json:"providers,omitempty" yaml:"providers,omitempty"`
}

// RetryConfig holds retry configuration for stage execution.
type RetryConfig struct {
	// MaxAttempts is the maximum retry attempts (default: 3).
	MaxAttempts uint32 `json:"max_attempts,omitempty" yaml:"max_attempts,omitempty"`

	// BackoffMs is the initial backoff delay in milliseconds (default: 1000).
	BackoffMs uint64 `json:"backoff_ms,omitempty" yaml:"backoff_ms,omitempty"`

	// BackoffMultiplier is the exponential backoff multiplier (default: 2.0).
	BackoffMultiplier float64 `json:"backoff_multiplier,omitempty" yaml:"backoff_multiplier,omitempty"`
}

// DefaultRetryConfig returns a RetryConfig with default values.
func DefaultRetryConfig() RetryConfig {
	return RetryConfig{
		MaxAttempts:       3,
		BackoffMs:         1000,
		BackoffMultiplier: 2.0,
	}
}

// Stage represents a single stage in the pipeline.
type Stage struct {
	// ID is the stage identifier (unique within pipeline).
	ID string `json:"id" yaml:"id"`

	// Name is the human-readable stage name.
	Name string `json:"name" yaml:"name"`

	// Type is the stage type discriminator.
	Type StageType `json:"type" yaml:"type"`

	// UpstreamStageIDs are IDs of stages this depends on (DAG edges).
	UpstreamStageIDs []string `json:"upstream_stage_ids,omitempty" yaml:"upstream_stage_ids,omitempty"`

	// LlmConfig is the LLM configuration (required for llm and hybrid stages).
	LlmConfig *LlmStageConfig `json:"llm_config,omitempty" yaml:"llm_config,omitempty"`

	// ClipsConfig is the CLIPS configuration (required for clips_eval, clips_gen, and hybrid stages).
	ClipsConfig *ClipsConfig `json:"clips_config,omitempty" yaml:"clips_config,omitempty"`

	// TimeoutMs is the stage execution timeout in milliseconds (default: 30000).
	TimeoutMs uint64 `json:"timeout_ms,omitempty" yaml:"timeout_ms,omitempty"`

	// Retry is the retry configuration.
	Retry *RetryConfig `json:"retry,omitempty" yaml:"retry,omitempty"`
}

// PipelineDefinition represents a complete pipeline configuration.
type PipelineDefinition struct {
	// ID is the unique pipeline identifier (UUID).
	ID string `json:"id" yaml:"id"`

	// Name is the human-readable pipeline name.
	Name string `json:"name" yaml:"name"`

	// Description is an optional pipeline description.
	Description string `json:"description,omitempty" yaml:"description,omitempty"`

	// Version is the schema version (default: "1.0").
	Version string `json:"version,omitempty" yaml:"version,omitempty"`

	// Stages is the ordered list of pipeline stages.
	Stages []Stage `json:"stages" yaml:"stages"`

	// CreatedAt is the creation timestamp (ISO 8601).
	CreatedAt time.Time `json:"created_at" yaml:"created_at"`

	// UpdatedAt is the last modification timestamp.
	UpdatedAt time.Time `json:"updated_at" yaml:"updated_at"`

	// Metadata contains arbitrary key-value metadata.
	Metadata map[string]any `json:"metadata,omitempty" yaml:"metadata,omitempty"`
}

// MultiProviderResult holds results from multi-provider stage execution.
type MultiProviderResult struct {
	// Results keyed by provider name.
	Results map[string]json.RawMessage `json:"results"`
}

// GetDefaultTimeoutMs returns the default stage timeout (30000ms).
func GetDefaultTimeoutMs() uint64 {
	return 30000
}

// GetDefaultVersion returns the default schema version ("1.0").
func GetDefaultVersion() string {
	return "1.0"
}

File: packages/nxuskit-engine/crates/nxuskit-engine/src/pipeline/definition.rs
//! Pipeline definition types.
//!
//! This module defines the core types for pipeline configurations including
//! stages, configs, and the pipeline definition itself.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Stage type discriminator.
///
/// Determines what kind of processing a stage performs.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StageType {
    /// LLM-only stage (calls an LLM provider)
    Llm,
    /// CLIPS evaluation stage (runs CLIPS rules)
    ClipsEval,
    /// CLIPS generation stage (generates facts from CLIPS)
    ClipsGen,
    /// Hybrid stage (combines LLM and CLIPS)
    Hybrid,
}

impl std::fmt::Display for StageType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            StageType::Llm => write!(f, "llm"),
            StageType::ClipsEval => write!(f, "clips_eval"),
            StageType::ClipsGen => write!(f, "clips_gen"),
            StageType::Hybrid => write!(f, "hybrid"),
        }
    }
}

/// Source of CLIPS rules.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum RulesSource {
    /// Rules provided inline in the configuration
    Inline,
    /// Rules loaded from a file path
    File,
    /// Rules generated dynamically at runtime
    Dynamic,
}

/// Security validation severity level.
///
/// Controls how dangerous CLIPS constructs are handled.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum ValidationSeverity {
    /// Reject rules with dangerous constructs (default)
    #[default]
    Error,
    /// Log warning but proceed with loading
    Warning,
    /// Log info message only
    Info,
    /// Skip validation entirely
    Ignore,
}

/// Configuration for CLIPS-based stages.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ClipsConfig {
    /// Source of CLIPS rules
    pub rules_source: RulesSource,

    /// Inline rules or file path (based on rules_source)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rules_content: Option<String>,

    /// Persist rules after stage execution
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub save_rules: bool,

    /// Path for saved rules (required if save_rules is true)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub save_path: Option<String>,

    /// Use CLIPS binary format for save
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub bsave_binary: bool,

    /// Security validation severity level
    #[serde(default, skip_serializing_if = "is_default_severity")]
    pub validation_severity: ValidationSeverity,

    /// CLIPS execution timeout in milliseconds
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timeout_ms: Option<u64>,
}

fn is_default_severity(severity: &ValidationSeverity) -> bool {
    *severity == ValidationSeverity::Error
}

/// Configuration for LLM-based stages.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LlmStageConfig {
    /// Provider name (claude, openai, ollama, etc.)
    pub provider: String,

    /// Model identifier
    pub model: String,

    /// System prompt template
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system_prompt: Option<String>,

    /// User prompt template with {{variable}} placeholders
    pub user_prompt: String,

    /// Sampling temperature
    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f64>,

    /// Maximum response tokens
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_tokens: Option<u32>,

    /// Provider-specific parameters
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub additional_params: HashMap<String, serde_json::Value>,

    /// Multiple providers for parallel execution (optional)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub providers: Option<Vec<String>>,
}

/// Retry configuration for stage execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetryConfig {
    /// Maximum retry attempts
    #[serde(default = "default_max_attempts")]
    pub max_attempts: u32,

    /// Initial backoff delay in milliseconds
    #[serde(default = "default_backoff_ms")]
    pub backoff_ms: u64,

    /// Exponential backoff multiplier
    #[serde(default = "default_backoff_multiplier")]
    pub backoff_multiplier: f64,
}

fn default_max_attempts() -> u32 {
    3
}

fn default_backoff_ms() -> u64 {
    1000
}

fn default_backoff_multiplier() -> f64 {
    2.0
}

impl Default for RetryConfig {
    fn default() -> Self {
        Self {
            max_attempts: default_max_attempts(),
            backoff_ms: default_backoff_ms(),
            backoff_multiplier: default_backoff_multiplier(),
        }
    }
}

/// A single stage in the pipeline.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Stage {
    /// Stage identifier (unique within pipeline)
    pub id: String,

    /// Human-readable stage name
    pub name: String,

    /// Stage type discriminator
    #[serde(rename = "type")]
    pub stage_type: StageType,

    /// IDs of stages this depends on (DAG edges)
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub upstream_stage_ids: Vec<String>,

    /// LLM configuration (required for llm and hybrid stages)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub llm_config: Option<LlmStageConfig>,

    /// CLIPS configuration (required for clips_eval, clips_gen, and hybrid stages)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub clips_config: Option<ClipsConfig>,

    /// Stage execution timeout in milliseconds
    #[serde(
        default = "default_timeout_ms",
        skip_serializing_if = "is_default_timeout"
    )]
    pub timeout_ms: u64,

    /// Retry configuration
    #[serde(skip_serializing_if = "Option::is_none")]
    pub retry: Option<RetryConfig>,
}

fn default_timeout_ms() -> u64 {
    30000
}

fn is_default_timeout(timeout: &u64) -> bool {
    *timeout == 30000
}

/// A complete pipeline definition.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PipelineDefinition {
    /// Unique pipeline identifier (UUID)
    pub id: String,

    /// Human-readable pipeline name
    pub name: String,

    /// Optional pipeline description
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    /// Schema version
    #[serde(default = "default_version")]
    pub version: String,

    /// Ordered list of pipeline stages
    pub stages: Vec<Stage>,

    /// Creation timestamp (ISO 8601)
    pub created_at: String,

    /// Last modification timestamp
    pub updated_at: String,

    /// Arbitrary key-value metadata
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub metadata: HashMap<String, serde_json::Value>,
}

fn default_version() -> String {
    "1.0".to_string()
}

/// Result from multi-provider stage execution.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MultiProviderResult {
    /// Results keyed by provider name
    pub results: HashMap<String, serde_json::Value>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_stage_type_display() {
        assert_eq!(StageType::Llm.to_string(), "llm");
        assert_eq!(StageType::ClipsEval.to_string(), "clips_eval");
        assert_eq!(StageType::ClipsGen.to_string(), "clips_gen");
        assert_eq!(StageType::Hybrid.to_string(), "hybrid");
    }

    #[test]
    fn test_validation_severity_default() {
        assert_eq!(ValidationSeverity::default(), ValidationSeverity::Error);
    }

    #[test]
    fn test_retry_config_default() {
        let config = RetryConfig::default();
        assert_eq!(config.max_attempts, 3);
        assert_eq!(config.backoff_ms, 1000);
        assert_eq!(config.backoff_multiplier, 2.0);
    }
}
